File: src/p0021/golang/solution.go
/**
 * Merge two sorted linked lists and return it as a sorted list. The list should be made by splicing together the nodes of the first two lists.
 * 
 * Example 1:
 * 
 * Input: l1 = [1,2,4], l2 = [1,3,4]
 * Output: [1,1,2,3,4,4]
 * Example 2:
 * 
 * Input: l1 = [], l2 = []
 * Output: []
 * Example 3:
 * 
 * Input: l1 = [], l2 = [0]
 * Output: [0]
 *  
 * 
 * Constraints:
 * 
 * The number of nodes in both lists is in the range [0, 50].
 * -100 <= Node.val <= 100
 * Both l1 and l2 are sorted in non-decreasing order
 */


func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil && l2 == nil {
		return nil
	} else if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	}
	var ans, p *ListNode
	p1, p2 := l1, l2
	for p1 != nil && p2 != nil {
		if p == nil {
			ans = &ListNode{}
			p = ans
		} else {
			p.Next = &ListNode{}
			p = p.Next
		}
		if p1.Val > p2.Val {
			p.Val = p2.Val
			p2 = p2.Next
		} else {
			p.Val = p1.Val
			p1 = p1.Next
		}
	}
	for p1 != nil {
		p.Next = &ListNode{}
		p = p.Next
		p.Val = p1.Val
		p1 = p1.Next
	}
	for p2 != nil {
		p.Next = &ListNode{}
		p = p.Next
		p.Val = p2.Val
		p2 = p2.Next
	}
	return ans
}

File: src/p0021/python/solution.py
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if l1 == None and l2 == None:
            return None
        p1, p2 = l1, l2
        temp = list()
        while p1 != None and p2 != None:
            if p1.val < p2.val:
                temp.append(p1.val)
                p1 = p1.next
            else:
                temp.append(p2.val)
                p2 = p2.next
        while p1 != None:
            temp.append(p1.val)
            p1 = p1.next
        while p2 != None:
            temp.append(p2.val)
            p2 = p2.next
        ans = ListNode(temp[0])
        p = ans
        for i in range(1, len(temp)):
            p.next = ListNode(temp[i])
            p = p.next
        
        return ans